src/core/lwt.ml

```diff
@@ -2534,9 +2534,9 @@ struct
         let callback in_completion_loop _result =
           let State_may_now_be_pending_proxy p = may_now_be_proxy p in
           let p = underlying p in
-          List.iter cancel ps;
           let State_may_have_changed p =
             finish_nchoose_or_npick_after_pending in_completion_loop p [] ps in
+          List.iter cancel ps;
```

Comment:
I believe this reordering produces the correct behavior.

There is one weird thing about Lwt now, though:

- The [callback of `Lwt.pick`](https://github.com/ocsigen/lwt/blob/424e4c50ef84d359bf0389e9fdb37ef66cbd1c3f/src/core/lwt.ml#L2398-L2404) cancels the pending promises *before* completing its own ("final") promise.
- This modified callback of `Lwt.npick` cancels the pending promises *after* completing its own promise.

This means that code running in `on_cancel` and regular callbacks of the respective pending promises can observe different states of the final promise depending on which function is used. Likewise, code running in the callbacks of the final promises can observe different states of the (to-be) canceled promises, depending on which function is used.

I'm not sure if that's really a problem, since we don't document or guarantee either behavior, and I'm not aware of any pathological interactions due to picking one or the other, so one could argue that this doesn't need to be addressed right now.

However, it could cause problems for users refactoring their code. For completeness, this could be addressed by factoring out the [implicit final promise completion](https://github.com/ocsigen/lwt/blob/424e4c50ef84d359bf0389e9fdb37ef66cbd1c3f/src/core/lwt.ml#L2433) from `finish_nchoose_or_npick_after_pending`, so that function only collects the result, and the [callback](https://github.com/ocsigen/lwt/blob/424e4c50ef84d359bf0389e9fdb37ef66cbd1c3f/src/core/lwt.ml#L2534-L2541) can fully control the order of (1) figuring out the result (2) canceling the pending promises (3) setting the state of the final promise (the one returned by `npick` itself). Callbacks in both `npick` and `nchoose` will have to be modified to call `complete` explicitly.